### ISSUE TITLE
Datastore Schema Migration Feature

### ISSUE THREAD

**Issue by Alos** [spring-cloud/spring-cloud-gcp#1902]
It's typical for users to want to update their models and change the "schema" of the data stored. Currently adding new (primitive) values to an entity already stored will cause the application to throw an exception. For example:

```Java
@Entity
public class Dog{
   @Id private Long id;
   private String property1;
   private String property2;
}
```

Store a few Dogs on your database and then modify the Entity:

```Java
@Entity
public class Dog {
   @Id private Long id;
   private String property1;
   private String property2;

  private boolean newField; //newly added field
}
```

Fetching one of these objects from the datastore will cause an NPE:

```
java.lang.NullPointerException: null
	at java.base/sun.invoke.util.ValueConversions.unboxBoolean(ValueConversions.java:108) ~[na:na]
```

A potential solution is to create new annotations to allow users to specify if they want to keep primitive types to their defaults, ignore them, throw an exception or perform some additional actions.  Objectify has a nice [selection of such annotations](https://github.com/objectify/objectify/wiki/SchemaMigration) that we would want to have in Spring Cloud GCP.

**Comment by meltsufin** [comment]
If this was a traditional relational database with an actual schema, the problem would be that the new column would be storing `NULL` values for the existing rows, and it would be incorrect to convert `NULL`s to a boolean value on load.
Since, in Datastore, there is no strict schema, when you add a new property, the corresponding "column" for those rows simply doesn't exist in Datastore. So, we can choose what to do in that situation. It might be valid to just not populate the object properties that don't have a corresponding "column" in the database. However, they will be populated wihen the object is written back to the database. So, that might be a bit surprising.

**Comment by dmitry-s** [comment]
I think we can add annotations for default values. That shouldn't be too hard.

**Comment by dmitry-s** [comment]
fixed in #2505
